Support virtual threads in JVMTI Resume* and Suspend* functions

babsingh opened the pull request:
This PR impacts the following JVMTI functions:

- [Resume|Suspend]Thread
- [Resume|Suspend]ThreadList
- [Resume|Suspend]AllVirtualThreads

The above functions have been either implemented or updated as per the
JVMTI doc:
https://download.java.net/java/early_access/jdk19/docs/specs/jvmti.html

Other changes:
- Added boolean inspectingLiveVirtualThreadList to synchronize between
[Resume|Suspend]AllVirtualThreads, JVM_VirtualThreadMountBegin-End and
JVM_VirtualThreadUnmountBegin-End.
- If a thread is suspended while waiting on the monitor, then the
monitor is released after waking up and before invoking
internalEnterVMFromJNI, where the thread will be halted until resumed.
After resuming, the monitor is acquired again. This helps avoid
deadlocks.
- Added hidden field isSuspendedByJVMTI in VirtualThread to track if the
thread was suspended in JVMTI. If isSuspendedByJVMTI is non-zero, then
J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND is set in carrier J9VMThread's
publicFlags while resetting isSuspendedByJVMTI to zero. During mount,
this suspends the thread if the thread was unmounted while JVMTI
suspended it.

Related: #15760

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh commented:
Encountering hangs; debugging on my end. @fengxue-IS Let me know if you see anything out-of-order in these changes.

fengxue-IS commented on runtime/jvmti/suspendhelper.cpp:
```diff
@@ -35,33 +35,49 @@ suspendThread(J9VMThread *currentThread, jthread thread, UDATA allowNull, UDATA
 			}
 		}
+#if JAVA_SPEC_VERSION >= 19
+		if (isVirtual) {
```
this should be an `else if` since mounted virtual thread will be suspended by the public flag, adding an extra suspend to the vthread state will cause logic issue?

fengxue-IS replied:
also, virtual thread should also check if the thread is alive or not and return `rc = JVMTI_ERROR_THREAD_NOT_ALIVE;` in those cases

fengxue-IS replied:
the for mounted continuation, `getVirtualThreadState` should return the vmthread state which should contain `JVMTI_THREAD_STATE_SUSPENDED` https://github.com/eclipse-openj9/openj9/blob/master/runtime/util/thrinfo.c#L409-L415

fengxue-IS commented:
Are you seeing the hang on just suspend/resume or suspendAll/ResumeAll ?

babsingh commented:
> Are you seeing the hang on just suspend/resume or suspendAll/ResumeAll ?

`suspendAll/ResumeAll` at the below location:
https://github.com/eclipse-openj9/openj9/blob/ef6948124a16e50c03faa7e69383cb9677bc985d/runtime/jvmti/suspendhelper.cpp#L53

keithc-ca commented on runtime/jvmti/jvmtiThread.c:
```diff
@@ -1072,11 +1078,30 @@ resumeThread(J9VMThread * currentThread, jthread thread)
+			isVirtual = IS_VIRTUAL_THREAD(currentThread, J9_JNI_UNWRAP_REFERENCE(thread));
+		}
+		if (isVirtual) {
+			/* targetThread is NULL only for an unmounted virtual threads. */
```
grammar: please remove "an"

babsingh replied:
None of the existing changes may stay. Will make sure it is fixed in the finalized changes if the comment stays.

babsingh commented:
The following tests have successfully passed:
- https://github.com/ibmruntimes/openj9-openjdk-jdk19/tree/openj9/test/hotspot/jtreg/serviceability/jvmti/vthread/SuspendResume2
- https://github.com/ibmruntimes/openj9-openjdk-jdk19/tree/openj9/test/hotspot/jtreg/serviceability/jvmti/thread/GetThreadState

Note: Testing had to be run with `-Xint -Xgcpolicy:nogc` since segfaults originating from the GC and hangs from the JIT were seen.

The following tests could not be run successfully since some of the functions haven't been implemented:
- https://github.com/ibmruntimes/openj9-openjdk-jdk19/tree/openj9/test/hotspot/jtreg/serviceability/jvmti/vthread/SuspendResume1 [GetThreadListStackTraces has not been updated for vthreads yet] 
- https://github.com/ibmruntimes/openj9-openjdk-jdk19/tree/openj9/test/hotspot/jtreg/serviceability/jvmti/vthread/SuspendResumeAll [Issue with deriving carrier threads]

babsingh commented:
@keithc-ca All of your feedback has been addressed. Also, removed the draft state. Can you please complete your review?

keithc-ca commented on runtime/jvmti/jvmtiThread.c:
```diff
@@ -1072,12 +1078,29 @@ resumeThread(J9VMThread * currentThread, jthread thread)
+				rc = JVMTI_ERROR_THREAD_NOT_SUSPENDED;
+			} else {
+				J9VMJAVALANGVIRTUALTHREAD_SET_STATE(currentThread, threadObject, vthreadState & ~JVMTI_VTHREAD_STATE_SUSPENDED);
+			}
```
I suggest we should have an `else` block so we don't return `JVMTI_ERROR_NONE` if the thread is not virtual.

babsingh replied:
getVMThread has an assert which guarantees that targetThread is NULL only for virtual threads when mustBeAlive is TRUE. The `else if` can just be changed to an `else`.
https://github.com/eclipse-openj9/openj9/blob/f06248f89427d0509360ec6a1df74e0c06436072/runtime/jvmti/jvmtiHelpers.c#L178

keithc-ca commented on runtime/jvmti/jvmtiThread.c:
```diff
@@ -1248,22 +1271,141 @@ jvmtiGetCurrentThread(jvmtiEnv* env,
+jvmtiSuspendAllVirtualThreads(jvmtiEnv *env,
 	jint except_count,
-	const jthread* except_list)
+	const jthread *except_list)
```
If you're going to fix these, please fix them all (see e.g. lines 28, 30, 34-37, etc.).

babsingh replied:
They are being fixed in https://github.com/eclipse-openj9/openj9/pull/15914/files.

keithc-ca commented on runtime/jvmti/jvmtiThread.c:
```diff
@@ -1248,22 +1271,141 @@ jvmtiGetCurrentThread(jvmtiEnv* env,
+					}
+				}
+				if (suspend) {
+					UDATA currentThreadSuspended = 0;
```
This should use `FALSE` instead of `0` for consistency.

babsingh replied:
Updated.

keithc-ca commented on runtime/jvmti/jvmtiHelpers.c:
```diff
@@ -789,69 +789,71 @@ getVirtualThreadState(J9VMThread *currentThread, jthread thread)
+			vm->internalVMFunctions->resumeThreadForInspection(currentThread, targetThread);
+		} else {
+			j9object_t vThreadObject = J9_JNI_UNWRAP_REFERENCE(thread);
+			jint vThreadState = (jint) J9VMJAVALANGVIRTUALTHREAD_STATE(currentThread, vThreadObject);
```
Other places don't assume `NULL == targetThread` signals a virtual thread: why is that assumption safe here?

babsingh replied:
getVirtualThreadState is only called from one location, which verifies that the input thread parameter is a virtual thread instance.
https://github.com/eclipse-openj9/openj9/blob/f06248f89427d0509360ec6a1df74e0c06436072/runtime/jvmti/jvmtiThread.c#L71-L76

keithc-ca commented on runtime/jvmti/suspendhelper.cpp:
```diff
@@ -35,33 +35,48 @@ suspendThread(J9VMThread *currentThread, jthread thread, UDATA allowNull, UDATA
+			} else {
+				J9VMJAVALANGVIRTUALTHREAD_SET_STATE(currentThread, threadObject, vthreadState | JVMTI_VTHREAD_STATE_SUSPENDED);
+			}
+		}
```
Error case here for non-virtual threads?

babsingh replied:
Same as https://github.com/eclipse-openj9/openj9/pull/15903#discussion_r976521831.

babsingh commented:
@keithc-ca I have addressed all the comments in the last feedback. But I will have to rework my changes because a deadlock was found while running the non-public cert tests last night. The rework will include adding a new lock. One lock will explicitly manage the vThreadInspectCount, and the other lock will explicitly handle the vThreadList. None of the locks will be held if there is a halt on VM Access during suspension. The purpose of using two locks is to avoid intermittent deadlocks and improve perf by reducing the critical section.

keithc-ca commented on runtime/jvmti/jvmtiThread.c:
```diff
@@ -219,13 +219,17 @@ jvmtiSuspendThread(jvmtiEnv* env,
 	rc = getCurrentVMThread(vm, &currentThread);
 	if (rc == JVMTI_ERROR_NONE) {
-		UDATA currentThreadSuspended;
+		BOOLEAN currentThreadSuspended;
```
A local variable should be initialized in its declaration. See also lines 264 & 277 (and others which trust you'll address in #15914).

babsingh replied:
All formatting comments are addressed in https://github.com/eclipse-openj9/openj9/pull/15945.

keithc-ca commented:
> I will have to rework my changes because a deadlock was found

I'll move this to the draft state until you've addressed that.

gacholio commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -253,29 +253,24 @@ JVM_VirtualThreadMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
 	if (firstMount) {
-		if (NULL == vm->liveVirtualThreadList) {
+		f_monitorEnter(vm->virtualThreadListMutex);
+		while (vm->inspectVirtualThreadList) {
```
Implicit NULL check

gacholio replied:
Nevermind not a pointer.

gacholio commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -5780,8 +5780,10 @@ typedef struct J9JavaVM {
+	j9object_t *virtualThreadList;
+	omrthread_monitor_t virtualThreadInspectorMutex;
+	omrthread_monitor_t virtualThreadListMutex;
+	volatile BOOLEAN inspectVirtualThreadList;
```
Should this be a counter?

babsingh replied:
Currently, for functional correctness, only one thread is allowed to resumeAll or suspendAll vthreads at any time. So, a boolean should suffice for this purpose. Counter is not required since multiple threads are not allowed to inspect the list simultaneously.

gacholio replied:
That's my point - this seems unnecessarily restrictive. There's no one else who wants to examine the list? Looks like this is also protecting modification of the list (as it must).

babsingh replied:
> no one else who wants to examine the list?

[Resume|Suspend]AllVirtualThreads can be invoked from different threads simultaneously. It is to protect vthreads from being resumed and suspended simultaneously, which will lead to undefined behaviour.

gacholio replied:
The use of mulitple mutexxes is also confusing? Why can't one be used to protect both the list and the flag/counter?

babsingh replied:
Will attempt to reimplement with one mutex; changing to draft.

gacholio commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -253,29 +253,24 @@ JVM_VirtualThreadMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
+			vmFuncs->internalExitVMToJNI(currentThread);
+			f_monitorWait(vm->virtualThreadListMutex);
+			vmFuncs->internalEnterVMFromJNI(currentThread);
+			threadObj = VM_VMHelpers::popObjectInSpecialFrame(currentThread);
```
If you want to save a little code, you could avoid the push/pop and just unwrap the JNI ref again here.

babsingh commented:
@gacholio All previous feedback has been addressed.
- Re-implemented with one monitor.
- VirtualThread.state can't be modified easily from JVMTI. For using this approach, it will require 1) invocation of Java methods from JVMTI methods in order to resume virtual thread, and 2) modification of VirtualThread's Java code to avoid race-conditions.
- Had to add a hidden field in VirtualThread as a work-around for the VirtualThread.state approach. This avoids all the above issues. Will need to your guidance on the hidden field: wanted the field to be boolean; but didn't know which macros to use for load and store.

gacholio commented:
Can you please detail the issue with the state field? Boolean fields (hidden or not) are 32 bits (no field is currently ever smaller than 32 bits).

gacholio commented:
I'm not sure your approach is correct - for normal threads, there is no distinction made between JVMTI or java suspending/resuming a thread. This makes the APIs completely undependable, but I suspect the new APIs are equally bad. You'll need to test/prove that JVMTI and java suspend for virtual threads either has the same issue or does not.

babsingh commented:
> test/prove that JVMTI and java suspend for virtual threads either has the same issue or does not.

Java resume and suspend throw UnsupportedOperationException if invoked on a virtual thread: https://download.java.net/java/early_access/loom/docs/api/java.base/java/lang/Thread.html#suspend(). The two approaches are incomparable.

> Can you please detail the issue with the state field? Boolean fields (hidden or not) are 32 bits (no field is currently ever smaller than 32 bits).

Just wanted to confirm the syntax and macros to use for a boolean field. Since boolean fields are 32 bits, the below usage is the correct approach?
```
vmFuncs->addHiddenInstanceField(vm, "java/lang/VirtualThread", "isSuspendedByJVMTI", "Z", &vm->isSuspendedByJVMTIOffset)

J9OBJECT_U32_STORE(currentThread, threadObject, vm->isSuspendedByJVMTIOffset, TRUE);

if (J9OBJECT_U32_LOAD(currentThread, threadObject, vm->isSuspendedByJVMTIOffset)) {
```

gacholio commented:
The macros are correct. You have not described the issue with the vthread state.

So there is no java way to suspend/resume a vthread?

babsingh commented:
> So there is no java way to suspend/resume a vthread?

- Internal (private) API in Java code can suspend a virtual thread during [VirtualThread.tryGetStackTrace](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/openj9/src/java.base/share/classes/java/lang/VirtualThread.java#L871) by setting the `SUSPENDED` flag in `VirtualThread.state`.
- Setting the `SUSPENDED` flag prevents [VirtualThread.runContinuation](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/4efd563f9c7560c3b69d6dd4fb77dfd27126fa1e/src/java.base/share/classes/java/lang/VirtualThread.java#L210) to run the continuation. The flag is set using a retry approach if the compare-and-set fails.

> the issue with the vthread state.

- After removing the `SUSPENDED` flag, [VirtualThread.submitRunContinuation](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/4efd563f9c7560c3b69d6dd4fb77dfd27126fa1e/src/java.base/share/classes/java/lang/VirtualThread.java#L888) needs to be invoked for rescheduling the thread. Otherwise, the thread won't run. Invoking [VirtualThread.submitRunContinuation](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/4efd563f9c7560c3b69d6dd4fb77dfd27126fa1e/src/java.base/share/classes/java/lang/VirtualThread.java#L888) or similar Java code from JVMTI to resume a thread will be inefficient.
- If the thread is waiting in `JVM_VirtualThreadMountBegin` while being suspended by JVMTI, then more logic will be required because [VirtualThread.submitRunContinuation](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/4efd563f9c7560c3b69d6dd4fb77dfd27126fa1e/src/java.base/share/classes/java/lang/VirtualThread.java#L888) or similar code should not be invoked to resume a thread in this case.
- Race condition: If JVMTI sets the SUSPENDED flag in `VirtualThread.state`, then [VirtualThread.tryGetStackTrace](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/openj9/src/java.base/share/classes/java/lang/VirtualThread.java#L871)  may unset it before JVMTI resumes the thread.
- The hidden field approach does not depend on `VirtualThread.state`, and it suspends at `JVM_VirtualThreadMountEnd`. This prevents any conflict with the Java approach for suspending a virtual thread.

gacholio commented:
Regarding the race condition above, won't that already occur if multiple threads are trying to get the stack trace, or is there synchronization to prevent this?

babsingh commented:
> Regarding the race condition above, won't that already occur if multiple threads are trying to get the stack trace, or is there synchronization to prevent this?

Jack says that the compare-and-set will guarantee only one thread is allowed.

babsingh commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -253,20 +253,32 @@ JVM_VirtualThreadMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
+		if (OMR_ARE_ANY_BITS_SET(currentThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND)) {
+			/* Release the monitor before suspending. */
+			f_monitorExit(vm->liveVirtualThreadListMutex);
+			releasedLiveVirtualThreadListMutex = TRUE;
```
Can you indicate the location where the monitor should be released and reacquired?

babsingh replied:
VMAccess and monitor usage are interlinked since we are reading from an object and the object field is synchronized by the monitor. Not sure if we can efficiently achieve the above suggestion.

gacholio replied:
I'll have to think about this a bit - intially I was thinking you just remove the "if suspended" check.

gacholio replied:
I think the pattern should be:
```
enter VM
acquire monitor
while (condition) {
   exit VM
   wait
   release monitor
   enter VM
   acquire monitor
}
```

babsingh replied:
Makes sense. This will protect us from misbehaviour in scenarios where `currentThread->publicFlags` is outdated. Trying it out.

babsingh replied:
Updated. Local testing with OpenJDK tests (SuspendResume2 and SuspendResumeAll) passes.

keithc-ca commented on runtime/jvmti/jvmtiHelpers.c:
```diff
@@ -875,6 +875,9 @@ getVirtualThreadState(J9VMThread *currentThread, jthread thread)
 				rc = JVMTI_ERROR_INTERNAL;
 			}
+			if (0 != J9OBJECT_U32_LOAD(currentThread, vThreadObject, vm->isSuspendedByJVMTIOffset)) {
+				rc  |= JVMTI_THREAD_STATE_SUSPENDED;
```
nit: extra space after `rc`

babsingh replied:
Fixed.

gacholio commented:
I just need to go over in detail the interaction between the suspend flag in the vthread and the halt flag to see if there are any timing holes.

gacholio commented on runtime/jvmti/suspendhelper.cpp:
```diff
@@ -39,33 +39,49 @@ suspendThread(J9VMThread *currentThread, jthread thread, BOOLEAN allowNull, BOOL
-					*currentThreadSuspended = TRUE;
+				if (OMR_ARE_ANY_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_STOPPED)) {
+					/* Do not suspend dead threads. Mirrors SUN behaviour. */
+					rc = JVMTI_ERROR_THREAD_NOT_ALIVE;
```
Have you verified this behaviour with the normal thread suspend? I don't see anything in the new JVMTI spec calling this out (this error is not listed as a possible return code).

gacholio replied:
Sorry, confused by the diff again - this only applies to real threads.

gacholio commented:
Are only unmounted vthreads marked as suspended using the new field?

gacholio commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -334,9 +340,19 @@ JVM_VirtualThreadMountEnd(JNIEnv *env, jobject thread, jboolean firstMount)
 
+	/* If isSuspendedByJVMTI is non-zero, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND is set
+	 * in currentThread->publicFlags while resetting isSuspendedByJVMTI to zero. During
+	 * mount, this suspends the thread if the thread was unmounted while JVMTI suspended it.
```
while -> when

gacholio commented on runtime/jvmti/jvmtiHelpers.c:
```diff
@@ -875,6 +875,9 @@ getVirtualThreadState(J9VMThread *currentThread, jthread thread)
 				Assert_JVMTI_unreachable();
 				rc = JVMTI_ERROR_INTERNAL;
 			}
+			if (0 != J9OBJECT_U32_LOAD(currentThread, vThreadObject, vm->isSuspendedByJVMTIOffset)) {
```
Is there a case missing here (after the vthread mount clears this flag and sets the publicFlag then blocks).

babsingh replied:
It should fall under the vthread mounted case, which is covered by the `if (NULL != targetThread)` block.

gacholio replied:
What is the sequence for mounting/unmounting (regarding the start/end/enter/yield natives).

gacholio commented on runtime/jvmti/jvmtiThread.c:
```diff
@@ -1300,18 +1317,160 @@ jvmtiSuspendAllVirtualThreads(jvmtiEnv *env,
+
+		for (i = 0; i < except_count; ++i) {
+			jthread thread = except_list[i];
+			if ((NULL == thread)
```
Strictly speaking, this check is incomplete - it's legal to have a JNI ref to NULL (think weak global ref). There's manu other occurrences of this error around, so we can address this in a future PR.

gacholio replied:
Nothing yet, so please open.

babsingh replied:
Opened: https://github.com/eclipse-openj9/openj9/issues/16081.

gacholio commented:
What is the justification for clearing the vthread flag when setting the publicFlag? Wouldn't things be simpler if the vhtread flag was always correct?

babsingh commented:
> Are only unmounted vthreads marked as suspended using the new field?

Yes. Only unmounted vthreads are marked as suspended using the new field. For mounted vthreads, the state is derived from the carrier thread's J9VMThread, and the new field is unused. 

> What is the justification for clearing the vthread flag when setting the publicFlag? Wouldn't things be simpler if the vhtread flag was always correct?

At this point, the vthread is successfully mounted. The vthread flag is cleared when setting the publicFlag to be consistent with how mounted and unmounted vthreads are currently marked as suspended. Setting the vthread flag for both mounted and unmounted vthreads will require an extra `isVirtualThread` check in `jvmtiThread.c::resumeThread` and `suspendhelper.cpp::suspendThread`.

gacholio commented:
@fengxue-IS Any comments?

gacholio commented:
jenkins test sanity win jdk8

gacholio commented:
jenkins test sanity,extended zlinux jdk19

fengxue-IS commented:
> @fengxue-IS Any comments?

nothing from me, I think everything have been address
